packages/server/graphql/traceGraphQL.ts

```diff
@@ -0,0 +1,361 @@
+/**
+ * @file Wrapping functions to instrument compiled queries.
+ * These conflict with the dd-trace graphql plugin since they use the same fields to store intermediate data
+ */
+import {Span, SpanOptions, TraceOptions, Tracer} from 'dd-trace'
+import {
+  DefinitionNode,
+  DocumentNode,
+  ExecutionResult,
+  getNamedType,
+  GraphQLField,
+  GraphQLFieldResolver,
+  GraphQLNamedType,
+  GraphQLObjectType,
+  GraphQLResolveInfo,
+  GraphQLSchema,
+  isObjectType,
+  OperationDefinitionNode
+} from 'graphql'
+import {Path} from 'graphql/jsutils/Path'
+import {CompiledQuery, compileQuery, CompilerOptions, isCompiledQuery} from 'graphql-jit'
+
+interface ExecutionArgs {
+  rootValue?: any
+  contextValue?: any
+  variableValues?: any
+}
+
+interface Config {
+  /**
+   * Whether to collapse list items into a single element. (i.e. single
+   * `users.*.name` span instead of `users.0.name`, `users.1.name`, etc)
+   *
+   * @default true
+   */
+  collapse?: boolean
+
+  /**
+   * The maximum depth of fields/resolvers to instrument. Set to `0` to only
+   * instrument the operation or to `-1` to instrument all fields/resolvers.
+   *
+   * @default -1
+   */
+  depth?: number
+
+  /**
+   * An object of optional callbacks to be executed during the respective
+   * phase of a GraphQL operation. Undefined callbacks default to a noop
+   * function.
+   *
+   * @default {}
+   */
+  hooks?: {
+    execute?: (span: Span, args: ExecutionArgs, res?: any) => void
+  }
+}
+
+/**
+ * Return {@link compileQuery} function which will add tracing to the query.
+ */
+export function tracedCompileQuery(tracer: Tracer, config: Config) {
+  if (process.env.DD_TRACE_ENABLED !== 'true') {
+    return compileQuery
+  }
+  return function compileTracedQuery(
+    schema: GraphQLSchema,
+    document: DocumentNode,
+    operationName?: string,
+    partialOptions?: Partial<CompilerOptions>
+  ) {
+    wrapSchema(tracer, config, schema)
```

Comment:
I like how you call `wrapSchema` here instead of in another file like `rootSchema`. Feels more modular.

Ideally we could avoid traversing the schema for every new compiled query by flagging the entire schema as "wrapped".

Additionally, if we could guarantee that each type and type field is visited exactly once, then we could remove the recursive calls and the `PatchedMarker` logic.

For your consideration:

```ts
let isSchemaWrapped = false
function wrapSchema(tracer: Tracer, config: Config, schema: GraphQLSchema) {
  if (isSchemaWrapped) return
  isSchemaWrapped = true
  const typeMap = schema.getTypeMap()
  Object.values(typeMap).forEach((namedType) => {
    // ignore introspection and scalar types
    if (namedType.name.startsWith('__') || !isObjectType(namedType)) return
    const fields = namedType.getFields()
    Object.values(fields).forEach((field) => {
      if (field.resolve) {
        field.resolve = wrappedResolve(tracer, config, field.resolve)
      }
    })
  })
}
```